packages/carbon-web-components/src/components/side-panel/side-panel.ts

```diff
@@ -0,0 +1,1085 @@
+
+    if (this._hasActionToolbar) {
+      for (let action of actions) {
+        // action size should always be md
+        action.setAttribute('size', 'sm');
+      }
+    }
+  }
+
+  @state()
+  _actionsStacked = false;
+
+  @state()
+  _actionsCount = 0;
+
+  private checkSetActionsStacked() {
+    let stacked =
+      /(xs)|(sm)/.test(this.size) ||
+      (/md/.test(this.size) && this._actionsCount > 2);
+
+    if (!stacked && window && this._actions) {
+      const styles = window.getComputedStyle(this._actions);
+      stacked = styles['flex-direction'] === 'column';
+    }
+
+    if (this._actionsStacked !== stacked) {
+      this._actionsStacked = stacked;
+      this.dispatchEvent(
+        new CustomEvent(
+          (this.constructor as typeof CDSSidePanel).eventActionsStackingChange,
+          {
+            detail: {
+              stacked: stacked,
+            },
+          }
+        )
+      );
+      // actions may stack automatically based on button set component
+      this._updateActionSizes();
+    }
+  }
+
+  private _handleActionsChange(e: Event) {
+    const target = e.target as HTMLSlotElement;
+    const actions = target?.assignedElements();
+
+    // // wraps target with a button set
+    // let wrapper = target.querySelector(`.${blockClass}__actions-container`);
+    // if (!wrapper) {
+    //   target.insertAdjacentHTML(
+    //     'afterbegin',
+    //     '<cds-button-set></cds-button-set>'
+    //   );
+
+    //   wrapper = target.children[0];
+    // }
+    // wrapper.classList.add(`${blockClass}__actions-container`);
+
+    // for (let i = 0; i < actions.length; i++) {
+    //   wrapper.appendChild(actions[i]);
+    // }
+    const actionsCount = actions?.length ?? 0;
+    if (actionsCount === 0) return;
+
+    this.checkSetActionsStacked();
+
+    for (let i = 0; i < actionsCount; i++) {
+      if (i > 3) {
+        // hide excessive side panel actions
+        actions[i].setAttribute('hidden', 'hidden');
+        actions[i].setAttribute(
+          'data-actions-limit-3-exceeded',
+          `${actions.length}`
+        );
+      } else {
+        actions[i].setAttribute('size', this.condensedActions ? 'lg' : 'xl');
+        actions[i].classList.add(`${blockClassActionSet}__action-button`);
+        if (i === 0) {
+          actions[i].classList.add(
+            `${blockClassActionSet}__action-button--first`
+          );
+        }
+      }
+
+      // tempB4.insertAdjacentElement('beforebegin', actions[i]);
+    }
+    // tempB4.remove();
+
+    if (actionsCount > 3) {
+      this._actionsCount = 3;
+      console.warn(`Too many side-panel actions, max 3.`);
+    } else {
+      this._actionsCount = actionsCount;
+    }
+  }
+
+  /**
+   * The `ResizeObserver` instance for observing element resizes for re-positioning floating menu position.
+   */
+  // TODO: Wait for `.d.ts` update to support `ResizeObserver`
+  // @ts-ignore
+  private _resizeObserver = new ResizeObserver(() => {
```

Comment:
Not sure why we need the resize observer for the action bar? Can you explain?

Reply:
Resize observer is top cope with the unknown height of the button set, which may depending on side panel width get stacked. I can probably work around its use with a container query (replicating code in ibm-products), but will have to experiment. Hopefully will not discover `:has` is needed as it's not currently in FF ESR. 

NOTE: the stylelint config is very old in the repo.